From fc8e1b30781b326100178554ad7bd5a2dade6adb Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Fri, 10 Oct 2014 21:03:19 -0400
Subject: [block] blk-mq: cleanup after blk_mq_init_rq_map failures

Message-id: <1412975015-5370-19-git-send-email-jmoyer@redhat.com>
Patchwork-id: 97463
O-Subject: [RHEL7 PATCH 18/34] blk-mq: cleanup after blk_mq_init_rq_map failures
Bugzilla: 1146660
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Mike Snitzer <snitzer@redhat.com>

This is a backport of the following commit.  This resolves bug
1146660.

  commit 5676e7b6db02b80eafc2e3ad316d5f2fee817ecb
  Author: Robert Elliott <elliott@hp.com>
  Date:   Tue Sep 2 11:38:44 2014 -0500

    blk-mq: cleanup after blk_mq_init_rq_map failures

    In blk-mq.c blk_mq_alloc_tag_set, if:
     set->tags = kmalloc_node()
    succeeds, but one of the blk_mq_init_rq_map() calls fails,
     goto out_unwind;
    needs to free set->tags so the caller is not obligated
    to do so.  None of the current callers (null_blk,
    virtio_blk, virtio_blk, or the forthcoming scsi-mq)
    do so.

    set->tags needs to be set to NULL after doing so,
    so other tag cleanup logic doesn't try to free
    a stale pointer later.  Also set it to NULL
    in blk_mq_free_tag_set.

    Tested with error injection on the forthcoming
    scsi-mq + hpsa combination.

    Signed-off-by: Robert Elliott <elliott@hp.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 4e8fe2a..098c936 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -1985,6 +1985,8 @@ int blk_mq_alloc_tag_set(struct blk_mq_tag_set *set)
 out_unwind:
  while (--i >= 0)
   blk_mq_free_rq_map(set, set->tags[i], i);
+ kfree(set->tags);
+ set->tags = NULL;
 out:
  return -ENOMEM;
 }
@@ -2000,6 +2002,7 @@ void blk_mq_free_tag_set(struct blk_mq_tag_set *set)
  }
 
  kfree(set->tags);
+ set->tags = NULL;
 }
 EXPORT_SYMBOL(blk_mq_free_tag_set);
 
-- 
1.7.1